src/psyclone/psyir/nodes/call.py

```diff
@@ -593,3 +610,188 @@ def _location_txt(node):
             f"Failed to find a Routine named '{rsym.name}' in "
             f"{_location_txt(root_node)}. This is normally because the routine"
             f" is within a CodeBlock.")
+
+    def _check_argument_type_matches(
+                self,
+                call_arg: DataSymbol,
+                routine_arg: DataSymbol,
+            ) -> bool:
+        """Return information whether argument types are matching.
+        This also supports 'optional' arguments by using
+        partial types.
+
+        :param call_arg: _description_
+        :type call_arg: DataSymbol
+        :param routine_arg: _description_
+        :type routine_arg: DataSymbol
+        :raises CallMatchingArgumentsNotFound: _description_
+        :raises CallMatchingArgumentsNotFound: _description_
+        """
+        if isinstance(
+            routine_arg.datatype, UnsupportedFortranType
+        ):
+            # This could be an 'optional' argument.
+            # This has at least a partial data type
+            if (
+                call_arg.datatype
+                != routine_arg.datatype.partial_datatype
+            ):
+                raise CallMatchingArgumentsNotFound(
+                    f"Argument partial type mismatch of call "
+                    f"argument '{call_arg}' and routine argument "
+                    f"'{routine_arg}'"
+                )
+        else:
+            if call_arg.datatype != routine_arg.datatype:
+                raise CallMatchingArgumentsNotFound(
+                    f"Argument type mismatch of call argument "
+                    f"'{call_arg}' and routine argument "
+                    f"'{routine_arg}'"
+                )
+
+        return True
+
+    def _get_argument_routine_match(self, routine: Routine):
+        '''Return a list of integers giving for each argument of the call
+        the index of the argument in argument_list (typically of a routine)
```

Comment:
"...of the corresponding entry in the argument list of the supplied routine."

Reply:
OK.